File: trees/bst_operations.py
class Node:
    def __init__(self, value):
        self.data = value
        self.left = None
        self.right = None

# Insert node
def insert(root, value):
    if root is None:
        return Node(value)
    if value < root.data:
        root.left = insert(root.left, value)
    elif value > root.data:
        root.right = insert(root.right, value)
    return root

# Find minimum value node
def min_value_node(node):
    current = node
    while current.left:
        current = current.left
    return current

# Delete node
def delete_node(root, value):
    if root is None:
        return root

    if value < root.data:
        root.left = delete_node(root.left, value)
    elif value > root.data:
        root.right = delete_node(root.right, value)
    else:
        if root.left is None:
            return root.right
        elif root.right is None:
            return root.left

        temp = min_value_node(root.right)
        root.data = temp.data
        root.right = delete_node(root.right, temp.data)

    return root

# Inorder traversal
def inorder(root):
    if root:
        inorder(root.left)
        print(root.data, end=" ")
        inorder(root.right)

# Driver code
root = None
values = [50, 30, 70, 20, 40, 60, 80]

for v in values:
    root = insert(root, v)

print("Inorder Traversal:", end=" ")
inorder(root)

root = delete_node(root, 20)
print("\nAfter Deletion:", end=" ")
inorder(root)

File: trees/BSTOperations.java
class Node {
    int data;
    Node left, right;

    Node(int value) {
        data = value;
        left = right = null;
    }
}

public class BSTOperations {
    Node root;

    // Insert node
    Node insert(Node root, int value) {
        if (root == null)
            return new Node(value);

        if (value < root.data)
            root.left = insert(root.left, value);
        else if (value > root.data)
            root.right = insert(root.right, value);

        return root;
    }

    // Find minimum value
    Node minValueNode(Node node) {
        Node current = node;
        while (current.left != null)
            current = current.left;
        return current;
    }

    // Delete node
    Node deleteNode(Node root, int value) {
        if (root == null)
            return root;

        if (value < root.data)
            root.left = deleteNode(root.left, value);
        else if (value > root.data)
            root.right = deleteNode(root.right, value);
        else {
            if (root.left == null)
                return root.right;
            else if (root.right == null)
                return root.left;

            Node temp = minValueNode(root.right);
            root.data = temp.data;
            root.right = deleteNode(root.right, temp.data);
        }
        return root;
    }

    // Inorder traversal
    void inorder(Node root) {
        if (root != null) {
            inorder(root.left);
            System.out.print(root.data + " ");
            inorder(root.right);
        }
    }

    public static void main(String[] args) {
        BSTOperations tree = new BSTOperations();

        tree.root = tree.insert(tree.root, 50);
        tree.insert(tree.root, 30);
        tree.insert(tree.root, 70);
        tree.insert(tree.root, 20);
        tree.insert(tree.root, 40);
        tree.insert(tree.root, 60);
        tree.insert(tree.root, 80);

        System.out.print("Inorder Traversal: ");
        tree.inorder(tree.root);

        tree.root = tree.deleteNode(tree.root, 20);
        System.out.print("\nAfter Deletion: ");
        tree.inorder(tree.root);
    }
}
